mount/upgrader.go

```diff
@@ -150,22 +161,46 @@ func (u *Upgrader) refetch(ctx context.Context) error {
 	return nil
 }
 
-// DeleteTransient deletes the transient associated with this Upgrader, if
-// one exists. It is the caller's responsibility to ensure the transient is
-// not in use.
-func (u *Upgrader) DeleteTransient() error {
-	u.lk.Lock()
-	defer u.lk.Unlock()
-
-	if u.transient == "" {
-		return nil // nothing to do.
+func (u *Upgrader) toTransientReaderCloserUnlocked() (*transientReaderCloser, error) {
+	f, err := os.Open(u.transientFilePath())
+	if err != nil {
+		return nil, err
 	}
+	u.refs++
 
-	err := os.Remove(u.transient)
-	if err != nil {
+	return &transientReaderCloser{
+		Reader: f,
+		u:      u,
+	}, nil
+}
+
+func (u *Upgrader) transientFilePath() string {
+	return filepath.Join(u.rootdir, u.shardKey.String())
+}
+
+type transientReaderCloser struct {
+	Reader
+	u *Upgrader
+}
+
+func (m *transientReaderCloser) Close() error {
```

Comment:
The reason we don't want to remove transients immediately when all acquirers have been closed is because this leads to inefficient space utilisation. Imagine you have 2TiB available for scratch space, and you receive 20 requests in a row (but not overlapping) for shard A, which weighs 64GiB. You will keep refetching and deleting those 64GiB over and over again, even if you had 97% unused space.

The transient management will probably work through an LRU/LFU cache with a size boundary that allows us to evict (delete) the least/eldest accessed shards on demand, as we need to reclaim space to serve new shards that we still don't have locally.

Reply:
@raulk Yeah , I understand that. I implemented this as better GC is something we've planned for the future.